Add more metrics to our cluster and user operators

scholzj opened the pull request:
### Type of change

- Enhancement / new feature

### Description

Our operators currently provides only the basic metrics for JVM, Vert.x etc. Thsi PR adds some additional metrics from the operator world to Cluster and User Operators:
* Number of custom resources seen by the operator
* Number of periodical reconciliations
* Number of started individual reconciliations
* Number of failed reocnciliations
* Number of successful reconciliations
* Number of reconciliations which failed because of lock
* Average and max reconciliation times

It also adds new dashabord for operators (see attachement).

Metrics for Topic Oeprator will be added in separate PR. This PR does only some initial preparation to collect them from the Topic Operator container and to show the information in the dashabord.

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Update CHANGELOG.md

-----

![Screenshot 2020-03-31 at 21 28 04](https://user-images.githubusercontent.com/5658439/78067398-f7543100-7396-11ea-8c7e-fb54aec43a19.png)
![Screenshot 2020-03-31 at 21 28 22](https://user-images.githubusercontent.com/5658439/78067412-fae7b800-7396-11ea-8f95-c27cbbf40a9f.png)

scholzj commented:
Anyone interested in metrics and moniroting - any feedback with regards to the dashabord or metrics would be welcomed.

ppatierno commented on operator-common/src/main/java/io/strimzi/operator/common/AbstractOperator.java:
```diff
@@ -59,10 +64,46 @@
+        // Setup metrics
+        metrics = BackendRegistries.getDefaultNow();
+        
+        periodicReconciliationsCounter = Counter.builder("strimzi.periodical.reconcilitions")
```
if we could assume that `strimzi.reconciliations` is a sort of root for the reconciliations related metrics, should the others be `strimzi.reconciliations.periodical` and not the other way around?

ppatierno commented on operator-common/src/main/java/io/strimzi/operator/common/AbstractOperator.java:
```diff
@@ -259,15 +303,25 @@ public void accept(KubernetesClientException e) {
+                failedReconciliationsCounter.increment();
                 log.warn("{}: Failed to reconcile {}", reconciliation, cause.getMessage());
             } else if (!(cause instanceof UnableToAcquireLockException)) {
+                failedReconciliationsCounter.increment();
```
isn't it enough one `failedReconciliationsCounter.increment();` call out of this `if` and `elseif` statements?

scholzj replied:
No, because I do not think we should report when the reconciliation fails to obtain a lock as a failure.

stanlyDoge commented on operator-common/src/main/java/io/strimzi/operator/common/AbstractOperator.java:
```diff
@@ -59,10 +64,46 @@
+                .tag("kind", kind())
+                .register(metrics);
+
+        reconciliationsCounter = Counter.builder("strimzi.reconcilitions")
```
Typo in multiple locations `s/reconcilitions/reconciliations`

scholzj commented:
@ppatierno @stanlyDoge This should now be ready for another review.

@tombentley I requested the review from Paolo and Stanislav since they worked more with metrics. But if you are interested feel freee to review as well.

scholzj commented:
@strimzi-ci run tests

scholzj commented:
@PaulRMellor There are some minor changes to docs as well here. So if you have some time it would be great if you could review it as well. Thanks.

PaulRMellor commented on documentation/modules/proc-metrics-deploying-prometheus.adoc:
```diff
@@ -29,7 +29,9 @@ On MacOS, use:
 [source,shell,subs="+quotes,attributes"]
 sed -i '' 's/namespace: .*/namespace: _my-namespace_/' prometheus.yaml
 
-. Edit the `ServiceMonitor` resource in `strimzi-service-monitor.yaml` to define Prometheus jobs that will scrape the metrics data.
```
Do we want to say something in the introduction to this procedure about service and pod metrics data? Examples?

PaulRMellor commented:
@scholzj 
Changes look good
maybe we need a couple of updates elsewhere:

- We probably need to update the Introduction to metrics to mention Operators
https://strimzi.io/docs/0.16.2/#assembly-metrics-setup-str

- Maybe we should refer to the Introducing metrics section from the Operators chapter (4) too.
Something like what we have in the Kafka config sections: 
https://strimzi.io/docs/0.16.2/#assembly-metrics-deployment-configuration-kafka

scholzj commented:
@strimzi-ci run tests

scholzj commented:
@strimzi-ci run tests

scholzj commented:
@PaulRMellor Ok, I did the improvements you mentioned:
* Added a mentiond to Introduction to metrics
* Added a new subchapter to Operators

Please have a look. Thanks.

PaulRMellor commented:
> @PaulRMellor Ok, I did the improvements you mentioned:
> 
> * Added a mentiond to Introduction to metrics
> * Added a new subchapter to Operators
> 
> Please have a look. Thanks.

Great. Looks good.

ppatierno commented on examples/metrics/grafana-dashboards/strimzi-operators.json:
```diff
@@ -0,0 +1,1705 @@
+      "timeFrom": null,
+      "timeRegions": [],
+      "timeShift": null,
+      "title": "Started Reconciliation per hour",
```
I assume it's the number of reconciliation started due to added, updated or deleted CRs. Maybe just "Started" is not enough to understand, at least for me. Something around "CRs triggered Reconciliation per hour" ... but I have not strong opinion, just that "Started" doesn't clarify the nature.

scholzj replied:
The are not just the reconciliations triggered by the CR - such as added or modified. There are also the periodical once. There are 5 counters in general:
* The periodical reconciliations are fired everytime the timer fires .... increment by 1 for a single periodical reconcilition regardless the number of resources.
* The started reconciliations covers both watch and periodical. But is per resource. So if you have 2 Kafka clusters, they each increment by one
* Successful are the started which ended ok
* Failed are the started whithc failed with error other than no lock
* And the locked are those which didn't got the lock (tracked separately because they are not really errors per se)

ppatierno replied:
Yeah, these all the first 5 graphs in the picture but focusing on the "started ...." it could be not clear compared to the periodical one (even because you said that it contains the periodical one but is per resource). Maybe we can exclude the periodical ones in this "started ... " metric ?

scholzj replied:
No, we cannot exclude it easily. It is started because the counter is increased at startup. Where as the successful / faled / locked are at the end. I think you also need to count all the reconciliations to be able to corelate the started against the successful/failed/locked.
